interfaces/builtin/utils.go

```diff
@@ -32,46 +32,52 @@ import (
 // The maximum number of Usb bInterfaceNumber.
 const UsbMaxInterfaces = 32
 
-// AppLabelExpr returns the specification of the apparmor label describing
-// all the apps bound to a given slot. The result has one of three forms,
+// labelExpr returns the specification of the apparmor label describing
+// given apps and hooks. The result has one of three forms,
 // depending on how apps are bound to the slot:
 //
 // - "snap.$snap_instance.$app" if there is exactly one app bound
-// - "snap.$snap_instance.{$app1,...$appN}" if there are some, but not all, apps bound
-// - "snap.$snap_instance.*" if all apps are bound to the slot
-func appLabelExpr(apps map[string]*snap.AppInfo, snap *snap.Info) string {
+// - "snap.$snap_instance.{$app1,...$appN, $hook1...$hookN}" if there are some, but not all, apps/hooks bound
+// - "snap.$snap_instance.*" if all apps/hook are bound to the plug or slot
+func labelExpr(apps map[string]*snap.AppInfo, hooks map[string]*snap.HookInfo, snap *snap.Info) string {
 	var buf bytes.Buffer
+
+	names := make([]string, 0, len(apps)+len(hooks))
+	for appName := range apps {
+		names = append(names, appName)
+	}
+	for hookName := range hooks {
+		names = append(names, fmt.Sprintf("hook.%s", hookName))
+	}
+	sort.Strings(names)
+
 	fmt.Fprintf(&buf, `"snap.%s.`, snap.InstanceName())
-	if len(apps) == 1 {
-		for appName := range apps {
-			buf.WriteString(appName)
-		}
-	} else if len(apps) == len(snap.Apps) {
+	if len(names) == 1 {
+		buf.WriteString(names[0])
+	} else if len(apps) == len(snap.Apps) && len(hooks) == len(snap.Hooks) {
 		buf.WriteByte('*')
-	} else {
-		appNames := make([]string, 0, len(apps))
-		for appName := range apps {
-			appNames = append(appNames, appName)
-		}
-		sort.Strings(appNames)
+	} else if len(names) > 0 {
 		buf.WriteByte('{')
-		for _, appName := range appNames {
-			buf.WriteString(appName)
+		for _, name := range names {
+			buf.WriteString(name)
 			buf.WriteByte(',')
 		}
+		// remove trailing comma
 		buf.Truncate(buf.Len() - 1)
 		buf.WriteByte('}')
-	}
+	} // else: len(names)==0, gives "snap.<name>." that doesn't match anything
```

Comment:
This feels a little bit tenuous. We shouldn't have (type: app) snaps without any apps or hooks... what is a valid case when this is 0? Seems we should return an error, or possibly the empty string and adjusting the logic. We should never end up in a case where we are writing out a profile that is "snap.blah.", so I'm a little concerned this might paper over a real problem.

Reply:
> We shouldn't have (type: app) snaps without any apps or hooks... what is a valid case when this is 0?

actually some content snaps (which have type app) are like that, no apps, no hooks. Maybe you are thinking a snap with plugs and slots ?

Reply:
Ah, that's true. @stolowski - what happens with these snaps with this PR applied?

Reply:
They will be handled by the earlier case:
```} else if len(apps) == len(snap.Apps) && len(hooks) == len(snap.Hooks) {
 		buf.WriteByte('*')
...
```
(where boths sides have 0 apps/slots).

This last case with len(names)==0 only kicks in with core snap and implicit slots/implicit hooks, see the discussion below under the test case for "nothing".

Reply:
To further clarify, this case (to my understanding and observation of actual spread test failure as I was working on this PR) can only happen with core snap's its implicit hooks/slots and interfaces that (erroneously to my understanding) create a snippet with `peer=(label=###SLOT_SECURITY_TAGS###)` on **classic** (one such example is modem-manager as we discussed earlier on IRC), apart from label=unconfined snippet. With the old code we would end up with `peer=(label=snap.core.*)` here.